src/urllib3/multipart/decoder.py

```diff
@@ -0,0 +1,165 @@
+"""Logic for parsing and decomposing a multipart response body."""
+import email.parser
+import typing
+
+from .. import _collections
+from .encoder import encode_with
+
+if typing.TYPE_CHECKING:
+    from .. import response as _response
+
+
+@typing.overload
+def _split_on_find(content: str, bound: str) -> typing.Tuple[str, str]:
+    ...
+
+
+@typing.overload
+def _split_on_find(content: bytes, bound: bytes) -> typing.Tuple[bytes, bytes]:
+    ...
+
+
+def _split_on_find(
+    content: typing.AnyStr, bound: typing.AnyStr
+) -> typing.Tuple[typing.AnyStr, typing.AnyStr]:
+    point = content.find(bound)
+    return content[:point], content[point + len(bound) :]
+
+
+class ImproperBodyPartContentError(Exception):
+    pass
+
+
+class NonMultipartContentTypeError(Exception):
+    pass
+
+
+def _header_parser(
+    headers: bytes, encoding: str
+) -> typing.Sequence[typing.Tuple[str, str]]:
+    string = headers.decode(encoding)
+    items = email.parser.HeaderParser().parsestr(string).items()
+    items = typing.cast(typing.List[typing.Tuple[str, str]], items)
+    return items
+
+
+class BodyPart:
+    """This provides an easy way to interact with a single part of the body.
+
+    BodyParts include the headers in the part of the body as well as the body
+    content as bytes and optional converted to text based on the provided
+    encoding.
+
+    The encoding may be overridden by specifying ``part.encoding = '...'``.
+    """
+
+    def __init__(self, content: bytes, encoding: str):
+        self.encoding = encoding
+        headers: typing.Dict[str, str] = {}
+        # Split into header section (if any) and the content
+        if b"\r\n\r\n" in content:
```

Comment:
Can we use partition and check the middle element of the returned 3-tuple to avoid doing two scans for this bytestring?